usagereporter: on-prem dial home

espadolini opened the pull request:
This PR implements the public changes to Teleport for the implementation of gravitational/cloud#3830. Proto changes depend on gravitational/cloud#4018.

Will be enabled by gravitational/teleport.e#1061.

espadolini committed:
sync prehog/v1alpha/teleport.proto from cloud

espadolini committed:
AggregatingUsageReporter

espadolini committed:
graceful stopping

espadolini committed:
alerts WIP

espadolini committed:
Fix prehog protos package name

espadolini committed:
Split reporter and submitter

espadolini committed:
batch_uuid

espadolini committed:
Naming

espadolini committed:
more naming

espadolini committed:
Move backend stuff in a "service" object

espadolini committed:
Return the correct err from gracefulstop

espadolini committed:
Tweak alert link

espadolini commented:
In draft because it's missing unit tests and because the changes to the prehog proto should land in `gravitational/cloud` first.

espadolini committed:
Better field comments

zmb3 commented on lib/service/service.go:
```diff
@@ -1824,7 +1825,7 @@ func (process *TeleportProcess) initAuthService() error {
 			warnOnErr(listener.Close(), log)
 		}
-		if payload == nil {
+		if ctx, _ := payload.(context.Context); ctx == nil {
```
```suggestion
		if ctx := payload.(context.Context); ctx == nil {
```

espadolini replied:
changed to use `payloadContext` in de047f3a0a

zmb3 commented on lib/usagereporter/teleport/aggregating/reporter.go:
```diff
@@ -0,0 +1,316 @@
+	wg sync.WaitGroup
+	// baseCtx is the parent context for all operations (backend operations
+	// being the most important).
+	baseCtx context.Context
```
How about a proper close API rather than stuffing multiple contexts in a struct like this?

zmb3 reviewed (commented):
This implementation seems a bit over complicated for the task at hand. I worry that we're somewhat mis-using a number of concurrency primitives (contexts, waitgroups, and muteness).

espadolini committed:
payloadContext rather than a manual type assertion

espadolini commented on lib/usagereporter/teleport/aggregating/reporter.go:
```diff
@@ -0,0 +1,316 @@
+
+		// if r.mu is locked then something has just called
+		// maybeFinalizeReportLocked anyway
+		if r.mu.TryLock() {
```
Always locking I guess?
```go
		r.mu.Lock() {
		if !r.stopped {
			r.maybeFinalizeReportsLocked()
		}
		r.mu.Unlock()
```

espadolini commented on lib/usagereporter/teleport/aggregating/reporter.go:
```diff
@@ -0,0 +1,316 @@
+	r.mu.Unlock()
+
+	if len(userActivity) > 0 {
+		// we're allowed to do this because we know that we'll only Wait on r.wg
+		// later on in this function's body; anywhere else we're not allowed to
+		// Add while r.stopped is true, as Wait can't be called concurrently
+		// with an Add from 0, and if we began shutting down we can't assume
+		// that the count is above 0
+		r.wg.Add(1)
```
We call `Add` in three points: right before the periodic finalizer, and in the two places right before we spawn `persistUserActivity`; there's two deferred `Done`s (in `runPeriodicFinalize` and `persistUserActivity`) and a single `Wait`.

We can simplify this further by not caring if there's leftover goroutines still running before returning from `GracefulStop` I suppose, since they'll just be stopping due to context closure anyway, but is that actually an improvement?

espadolini replied:
edit: we actually can't, we rely on the waitgroup to confirm the end of the goroutines to actually give enough time for the backend operations to finish, letting them die on their own only works if the shutdown context is actually too short

espadolini committed:
Rework reporter logic

espadolini commented:
@zmb3 a913ac50a4c322742530c9062ef4b38342042854 reworks the `Reporter` logic - it's a lot more straightforward now, PTAL.

espadolini committed:
Close the base context at the end of run

espadolini committed:
Use the count parameter in listUserActivityReports

espadolini committed:
Use ttl parameter in createUserActivityReportsLock

espadolini committed:
service tests

espadolini committed:
submitter test

zmb3 commented on lib/usagereporter/teleport/aggregating/reporter.go:
```diff
@@ -0,0 +1,279 @@
+func (r *Reporter) GracefulStop(ctx context.Context) error {
+	r.closingOnce.Do(func() { close(r.closing) })
+
+	select {
```
I don't understand the semantics here - what is the passed in context used for?

espadolini replied:
Documented in the interface in 91b75dc06280c0406ce34c81da44722698b5c93c:
```go
	// GracefulStop gracefully closes and runs any finalization needed by the
	// UsageReporter; operations can run as long as the context is alive, but
	// must terminate quickly (even losing data) if the context is closed.
	// Returns nil if operations have completed cleanly.
	GracefulStop(context.Context) error
```

espadolini replied:
In the case of the `aggregating.Reporter`, closing the context that's passed in will trigger the closing of the base context, and thus all closing operations ungracefully (and quickly) fail.

zmb3 commented on lib/usagereporter/teleport/aggregating/submitter.go:
```diff
@@ -0,0 +1,192 @@
+			"reports":       len(reports),
+			"oldest_report": reports[0].GetStartTime().AsTime(),
+			"newest_report": reports[len(reports)-1].GetStartTime().AsTime(),
+		}).Errorf("Failed to send usage reports.")
```
```suggestion
		}).Error("Failed to send usage reports.")
```

espadolini committed:
reporter test

espadolini committed:
Document GracefulStopper.GracefulStop

espadolini committed:
Fix errorf/warnf that don't need formatting

espadolini committed:
Add debug line on a successful report

espadolini committed:
fix license

espadolini committed:
sync prehog proto from cloud

espadolini committed:
Only send non-nil events through test channel

This fixes a hang/data race, as we effectively "send" a nil event
whenever the ticker hits

espadolini commented on lib/usagereporter/teleport/aggregating/submitter.go:
```diff
@@ -0,0 +1,192 @@
+			alertName,
+			alertMessage,
+			types.WithAlertLabel(types.AlertOnLogin, "yes"),
+			types.WithAlertLabel(types.AlertPermitAll, "yes"),
```
I went with the same alert visibility as the license warnings (albeit at `AlertSeverity_MEDIUM` rather than `HIGH`, so orange rather than red); should I reduce the visibility? I think the intention is to make it similar to the old-world Houston license warning, which - from what I can tell - was shown to everyone.

espadolini replied:
FWIW nowadays we have alert acknowledgement in place, so admins can just hide the alert once they are aware of the problem and are attempting to fix it.

espadolini committed:
Docs and variable names

espadolini committed:
Add some debug info to the lock and the alert

espadolini commented on lib/usagereporter/teleport/aggregating/service.go:
```diff
@@ -0,0 +1,152 @@
+func (r reportService) createUserActivityReportsLock(ctx context.Context, ttl time.Duration) error {
+	if _, err := r.b.Create(ctx, backend.Item{
+		Key:     backend.Key(userActivityReportsLock),
+		Value:   []byte("null"),
```
I made it a string containing time and host ID, both in the lock and in the description of the cluster alert (e206aa4).

espadolini replied:
(I thought about using a label for the cluster alert, but it would have to come with a bunch of caveats about not actually ever using it as a label for RBAC purposes as it wouldn't be trustable, so I went with the metadata description field instead)

espadolini commented on lib/usagereporter/teleport/aggregating/submitter.go:
```diff
@@ -0,0 +1,192 @@
+	Submitter UsageReportsSubmitter
+}
+
+func (cfg *SubmitterConfig) CheckAndSetDefaults() error {
```
added

espadolini committed:
Sync prehog protos from cloud

This includes a move to prehog.v1 for the new rpc and messages

espadolini committed:
mechanical: rename imports of prehog/v1alpha

This renames imports of ".../prehog/v1alpha" from prehogv1a to prehogv1,
and imports of ".../prehog/v1alpha/prehogv1alphaconnect" from prehogv1c
to prehogv1ac, to avoid confusion with imports of the newly added
".../prehog/v1" and ".../prehog/v1/prehogv1connect".

espadolini committed:
Merge remote-tracking branch 'origin/master' into espadolini/usagereporter-onprem

espadolini committed:
Merge remote-tracking branch 'origin/master' into espadolini/usagereporter-onprem

espadolini commented:
Sorry for the noise, the manual merge lost its correct parents

espadolini commented:
Waiting for gravitational/cloud#4018 to confirm the .proto changes

espadolini committed:
Merge remote-tracking branch 'origin/master' into espadolini/usagereporter-onprem
